### PR TITLE
Typing cleanup

### PR DESCRIPTION
**Pull request by shenjackyuanjie**
Just fixing Pyright/ruff's error whatever I found
review ready

### CODE REVIEW

**Comment by shenjackyuanjie** [comment]
just ……
merge when you have time, I'm still working on it though, but I'll make sure every commit works

**Comment by caffeinepills** [comment]
A few things:

No reason to block I001 on these modules.

The typing for `pyglet/__init__.py` is specified in the `__init__.pyi`. For inspections to work in PyCharm for the main modules, the ModuleTypes need to be as they were previously or inspection autocomplete will break.

For things like :
```py
        if not isinstance(other, _InlineElementGroup):
            return False
```
It's probably better to put this in an assert, so it can be optimized out. The `self.__class__ is other.__class__` will always remove anything else that isn't the same class, so that is just redundant.

**Review by pushfoo — CHANGES_REQUESTED** [review]
I'm concerned by the repeated noqa which is getting a bit verbose. Since ruff added file-scoped noqa (https://github.com/astral-sh/ruff/pull/2978), it seems good to deploy file-specific ignores for things like `lib` which is going to do some nasty things.

Definitely keep:
* `__all__` improvements
* added argument type annotations

**Comment by shenjackyuanjie** [comment]
> A few things:
> 
> No reason to block I001 on these modules.
> 
> The typing for `pyglet/__init__.py` is specified in the `__init__.pyi`. For inspections to work in PyCharm for the main modules, the ModuleTypes need to be as they were previously or inspection autocomplete will break.
> 
> For things like :
> 
> ```python
>         if not isinstance(other, _InlineElementGroup):
>             return False
> ```
> 
> It's best this is left out or put this in an assert, so it can be optimized out. The `self.__class__ is other.__class__` will always bail out of anything else that isn't the same class during the comparison, so that adds unnecessary checks.

the problem is, it is unsorted

image

I'll fix the `isinstance` check

File: pyglet/lib.py
[the remark this answers is not in the document]
Comment (shenjackyuanjie) [reply pyglet/lib.py:1]: doing so

**Comment by shenjackyuanjie** [comment]
I'll just stop here, let's start review

File: examples/text/advanced_font.py [hunk 1]
```diff
@@ -1,26 +1,24 @@
-
-"""Example of advanced font rendering features. Currently only supported on Windows."""
-
+"""Example of advanced font rendering features."""
+# ruff: noqa: E501
 import pyglet
-pyglet.options["advanced_font_features"] = True
-
-if pyglet.compat_platform != 'win32':
```
[the remark this answers is not in the document]
Comment (shenjackyuanjie) [reply examples/text/advanced_font.py:7]: it dose works on mac and linux ( with a proper font )

[the remark this answers is not in the document]
Comment (pushfoo) [reply examples/text/advanced_font.py:7]: That needs to be documented then, and you need to be specific about what "work" means

File: pyglet/text/__init__.py [hunk 2]
```diff
@@ -437,9 +438,10 @@ class HTMLLabel(DocumentLabel):
     """
 
     def __init__(self, text: str = "", location: Location | None = None,
-                 x: float = 0.0, y: float = 0.0, z: float = 0.0, width: int | None = None, height: int | None = None,
+                 x: float = 0.0, y: float = 0.0, z: float = 0.0,
+                 width: int | None = None, height: int | None = None,
                  anchor_x: AnchorX = "left", anchor_y: AnchorY = "baseline", rotation: float = 0.0,
-                 multiline: bool = False, dpi: float | None = None,
+                 multiline: bool = False, dpi: int | None = None,
```
[the remark this answers is not in the document]
Comment (shenjackyuanjie) [reply pyglet/text/__init__.py:444]: then I'll change then inner dpi type into float

File: tools/gengl.py [hunk 3]
```diff
@@ -1,6 +1,5 @@
-"""
-Generate gl.py and gl_compat.py
-See : tools/requirements.txt
+"""Generate gl.py and gl_compat.py
+See : tools/requirements.txt.
```
[the remark this answers is not in the document]
Comment (shenjackyuanjie) [reply tools/gengl.py:2]: well, I'd like to fix them 😆

File: pyglet/lib.py [hunk 4]
```diff
@@ -218,16 +221,16 @@ def find_library(self, path):
         return None
 
     @staticmethod
-    def load_framework(name):
+    def load_framework(name: str):  # noqa: ANN205
```
[the remark this answers is not in the document]
Comment (shenjackyuanjie) [reply pyglet/lib.py:224]: thanks

File: pyglet/__init__.py [hunk 5]
```diff
@@ -435,26 +439,26 @@ def __setattr__(self, name, value):
...
-    lib = _ModuleProxy("lib")  # type: ignore
-    math = _ModuleProxy("math")  # type: ignore
-    media = _ModuleProxy("media")  # type: ignore
-    model = _ModuleProxy("model")  # type: ignore
-    resource = _ModuleProxy("resource")  # type: ignore
-    sprite = _ModuleProxy("sprite")  # type: ignore
-    shapes = _ModuleProxy("shapes")  # type: ignore
-    text = _ModuleProxy("text")  # type: ignore
-    window = _ModuleProxy("window")  # type: ignore
+    app = _ModuleProxy("app")
```
[the remark this answers is not in the document]
Comment (shenjackyuanjie) [reply pyglet/__init__.py:442]: then I'll add some ruff rule for it

File: examples/text/advanced_font.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (shenjackyuanjie) [reply examples/text/advanced_font.py:7]: good point, let's just keep it as windows only 😂

[the remark this answers is not in the document]
Comment (pushfoo) [reply examples/text/advanced_font.py:7]: Maybe we could warn instead and use it as a quick check tool? I wrote something like that for shapes.

File: examples/media/soundspace/reader.py [hunk 6]
```diff
@@ -78,7 +78,7 @@
     def read(self, file):
         if not hasattr(file, 'read'):
             self.basedir = os.path.dirname(file)
-            file = open(file, 'rt')
+            file = open(file)
```
[the remark this answers is not in the document]
Comment (shenjackyuanjie) [reply examples/media/soundspace/reader.py:81]: nice codeQL

File: examples/text/advanced_font.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (shenjackyuanjie) [reply examples/text/advanced_font.py:7]: ah, warnings, gogogo!

File: examples/text/advanced_font.py [hunk 7]
```diff
@@ -1,26 +1,25 @@
...
+
+if os.name == "nt":
+    # On Windows, it's possible to change the font anti-aliasing mode.
+    # Uncomment the below lines to set the options:
+    #
+    from pyglet.font.directwrite import DirectWriteGlyphRenderer
+    D2D1_TEXT_ANTIALIAS_MODE_DEFAULT = 0
+    D2D1_TEXT_ANTIALIAS_MODE_CLEARTYPE = 1 # not working?
+    D2D1_TEXT_ANTIALIAS_MODE_GRAYSCALE = 2 # same as default
+    D2D1_TEXT_ANTIALIAS_MODE_ALIASED = 3 # looks worse than grayscale
```
[the remark this answers is not in the document]
Comment (shenjackyuanjie) [reply examples/text/advanced_font.py:16]: pass

File: examples/text/advanced_font.py [hunk 8]
```diff
@@ -1,26 +1,25 @@
...
+import os
+
+if os.name == "nt":
+    # On Windows, it's possible to change the font anti-aliasing mode.
+    # Uncomment the below lines to set the options:
+    #
+    from pyglet.font.directwrite import DirectWriteGlyphRenderer
+    D2D1_TEXT_ANTIALIAS_MODE_DEFAULT = 0
+    D2D1_TEXT_ANTIALIAS_MODE_CLEARTYPE = 1 # not working?
+    D2D1_TEXT_ANTIALIAS_MODE_GRAYSCALE = 2 # same as default
```
[the remark this answers is not in the document]
Comment (shenjackyuanjie) [reply examples/text/advanced_font.py:15]: pass

File: examples/text/advanced_font.py [hunk 9]
```diff
@@ -1,26 +1,25 @@
...
-# DirectWriteGlyphRenderer.antialias_mode = D2D1_TEXT_ANTIALIAS_MODE_ALIASED
+import os
+
+if os.name == "nt":
+    # On Windows, it's possible to change the font anti-aliasing mode.
+    # Uncomment the below lines to set the options:
+    #
+    from pyglet.font.directwrite import DirectWriteGlyphRenderer
+    D2D1_TEXT_ANTIALIAS_MODE_DEFAULT = 0
+    D2D1_TEXT_ANTIALIAS_MODE_CLEARTYPE = 1 # not working?
```
[the remark this answers is not in the document]
Comment (shenjackyuanjie) [reply examples/text/advanced_font.py:14]: pass

File: pyglet/__init__.py [hunk 10]
```diff
@@ -322,44 +323,44 @@
...
-            print(f'  {indent}args=({", ".join(args)})')
+            print(f'  {indent}args=({", ".join(args)})') # noqa: T201
         else:
             for argname in code.co_varnames[:code.co_argcount]:
                 try:
                     argvalue = _trace_repr(frame.f_locals[argname])
-                    print(f"  {indent}{argname}={argvalue}")
-                except:
+                    print(f"  {indent}{argname}={argvalue}") # noqa: T201
+                except: # noqa: S110, E722, PERF203
```
[the remark this answers is not in the document]
Comment (shenjackyuanjie) [reply pyglet/__init__.py:349]: it's fine 😆

**Comment by shenjackyuanjie** [comment]
OSError: exception: access violation reading 0x0000018953E6D000
FINE

**Comment by shenjackyuanjie** [comment]
@benmoran56  I think this PR is good to go, please take some time to review it

File: pyproject.toml [hunk 11]
```diff
@@ -24,22 +26,25 @@ extend-safe-fixes = ["UP007", "UP006"]
...
     "SIM108", # Simplify if/else is not always best.
-    "PTH",  # don't require pathlib. Add eventually?
-    "TRY",  # no triceratops.
-    "NPY",  # remove numpy
-    "Q000",  # double quotes
+    "PTH123", # allow to use `open` not `Path.open`
+    "TRY",    # no triceratops.
+    "NPY",    # remove numpy
+    "Q000",   # double quotes
+    "I001",   # Ignore import order
```
Comment (caffeinepills) [inline pyproject.toml:45]: This should not be globally ignored.

Comment (shenjackyuanjie) [reply pyproject.toml:45]: hmmm, you mean I001? or all of them
I add I001, other three are added by you and @pushfoo

**Comment by shenjackyuanjie** [comment]
why always `OSError: exception: access violation reading 0x0000020A6D59F000`

**Comment by pushfoo** [comment]
It's a windows-specific issue. Can you check the following?
1. Does it happen locally?
2. Was it mentioned in other branches, PRs, or issues?

If it was fixed elsewhere, you might just need to rebase.

**Comment by shenjackyuanjie** [comment]
> It's a windows-specific issue according to the test runners. Can you check the following?
> 
> 1. Does it happen locally?
> 2. Was it mentioned in other branches, PRs, or issues?

well, until now, you can actually see two test faild, and they were both caused by the same issue
to local, I can rarely meet this issue, but it always comes with some kind of graphic driver issue

File: pyglet/window/__init__.py [hunk 12]
```diff
@@ -107,7 +107,7 @@ def on_draw():
     from pyglet.gl import CanvasConfig, Config, Context
     from pyglet.text import Label
 
-_is_pyglet_doc_run = hasattr(sys, 'is_pyglet_doc_run') and sys.is_pyglet_doc_run
+_is_pyglet_doc_run = hasattr(sys, 'is_pyglet_doc_run') and sys.is_pyglet_doc_run # pyright: ignore reportAttributeAccessIssue
```
Comment (benmoran56) [inline pyglet/window/__init__.py:110]: Can you move this into pyproject.toml? 
We want to avoid tool-specific overrides in the main code.

**Review by benmoran56 — APPROVED** [review]

File: examples/text/advanced_font.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (caffeinepills) [reply examples/text/advanced_font.py:7]: This example was made specially to show off the directwrite rendering options, and emoji support which is windows only.

File: pyglet/canvas/__init__.py [hunk 13]
```diff
@@ -22,7 +22,7 @@
 by the application; see the documentation for :class:`Screen`.
 
 .. versionadded:: 1.2
-"""
+"""  # noqa: I002
```
Comment (caffeinepills) [inline pyglet/canvas/__init__.py:25]: Still not necessary.

File: pyglet/font/user.py [hunk 14]
```diff
@@ -107,7 +107,7 @@ class UserDefinedFontBase(base.Font):
 
     def __init__(
             self, name: str, default_char: str, size: int, ascent: int | None = None, descent: int | None = None,
-            bold: bool = False, italic: bool = False, stretch: bool = False, dpi: int = 96, locale: str | None = None,
+            bold: bool = False, italic: bool = False, stretch: bool = False, dpi: float = 96, locale: str | None = None,
```
Comment (caffeinepills) [inline pyglet/font/user.py:110]: Should be int I believe.

File: pyproject.toml [hunk 15]
```diff
@@ -24,22 +26,25 @@ extend-safe-fixes = ["UP007", "UP006"]
...
     "S101", # allow asserts.
-    "E731",  # Allow lambda
-    "E741",  # Allow short variable names.
-    "D102",  # Do not check public method docstrings (Cannot determine overridden docstrings)
-             # @typing.override not available until python 3.12 (PEP 698)
-    "PLR",  # Too many of too many checks.
-    "ANN101",  # Ignore self type annotation.
+    "E731", # Allow lambda
+    "E741", # Allow short variable names.
+    "D102", # Do not check public method docstrings (Cannot determine overridden docstrings)
```
Comment (caffeinepills) [inline pyproject.toml:34]: I think should not be globally overridden as we may miss docstrings in the future.

File: pyproject.toml (hunk 11 above)
Comment (caffeinepills) [reply pyproject.toml:45]: Yes i001 should not be globally set.

**Review by caffeinepills — COMMENTED** [review]
Almost there, a few other things.

File: pyglet/font/user.py (hunk 14 above)
Comment (caffeinepills) [reply pyglet/font/user.py:110]: Actually this should be fine.

File: pyproject.toml (hunk 15 above)
Comment (shenjackyuanjie) [reply pyproject.toml:34]: this is not added by me though ……

**Review by benmoran56 — APPROVED** [review]

**Pull request merged**